internal/pkg/peer/blocksprovider/bft_censorship_monitor.go

```diff
@@ -0,0 +1,277 @@
+			m.logger.Warningf("Failure while launching header receiver monitors: %s", err)
+		}
+
+		select {
+		case <-m.stopCh:
+			m.logger.Debug("Stopping to monitor progress")
+			m.errorCh <- &errStopping{message: "received a stop signal"}
+			return
+		case <-time.After(bftBlockCensorshipTimeout / 100):
+			if m.detectBlockCensorship() {
+				m.errorCh <- &errCensorship{message: fmt.Sprintf("block censorship detected, endpoint: %s", m.fetchSources[m.blockSourceIndex])}
+				m.logger.Debug("Block censorship detected, stopping to monitor progress")
+				m.Stop()
+				return
+			}
+		}
+	}
+}
+
+func (m *bftCensorshipMonitor) ErrorsChannel() <-chan error {
+	return m.errorCh
+}
+
+func (m *bftCensorshipMonitor) Stop() {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if m.stopFlag {
+		return
+	}
+
+	m.stopFlag = true
+	close(m.stopCh)
+	for _, hRcvMon := range m.hdrRcvMonitors {
+		if hRcvMon.headerReceiver != nil {
+			_ = hRcvMon.headerReceiver.CloseSend()
+		}
+	}
+}
+
+func (m *bftCensorshipMonitor) detectBlockCensorship() bool {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	now := time.Now()
+	lastBlockNumber, lastBlockTime := m.progressReporter.BlockProgress()
+	if m.suspicion {
+		// When there is a suspicion, we're waiting for one of two things:
+		// - either a new block arrives, with a number >= than the header that triggered the suspicion,
+		//   in which case we lower the suspicion, or
+		// - the timeout elapses, in which case we assert the suspicion.
+		if lastBlockNumber >= m.suspicionBlockNumber {
+			m.logger.Debugf("[%s] last block number: %d >= than suspicion header number: %d; suspicion disproved", m.chainID, lastBlockNumber, m.suspicionBlockNumber)
+			m.suspicion = false
+		} else if now.After(m.suspicionTime.Add(bftBlockCensorshipTimeout)) {
+			m.logger.Warningf("[%s] block censorship timeout (%s) expired since suspicion time: %s; last block number: %d < than suspicion header number: %d; suspicion asserted",
+				m.chainID, bftBlockCensorshipTimeout, m.suspicionTime, lastBlockNumber, m.suspicionBlockNumber)
+			m.suspicion = false
+			return true
+		}
+	}
+
+	if m.suspicion {
+		// When there is a pending suspicion, advancing headers cannot disprove it, only a new block.
+		// Thus, there is no point in checking the progress of headers now.
+		return false
+	}
+
+	var ahead []timeNumber
+	for ep, hRcvMon := range m.hdrRcvMonitors {
+		if hRcvMon.headerReceiver == nil {
+			continue
+		}
+		headerNum, headerTime, err := hRcvMon.headerReceiver.LastBlockNum()
+		if err != nil {
+			continue
+		}
+		if headerNum > lastBlockNumber {
+			m.logger.Debugf("[%s] header receiver: %s, is ahead of block receiver; header num=%d, time=%s; block num=%d, time=%s", m.chainID, ep, headerNum, headerTime, lastBlockNumber, lastBlockTime)
+			ahead = append(ahead, timeNumber{
+				t: headerTime,
+				n: headerNum,
+			})
+		}
+	}
+
+	if len(ahead) == 0 {
+		return false
+	}
+
+	sort.Slice(ahead, func(i, j int) bool {
```

Comment:
why do you sort this? You can just iterate over it and find the max/min no? 

I know the computation here is negligible, but this is just awkward to sort in order to get the minimum / maximum.

Reply:
done.